### DISCUSSION TITLE
Batch image input for preprocessing on txt2img

### DISCUSSION THREAD

**Discussion by YourFriendlyNeighborhoodMONKE** [Mikubill/sd-webui-controlnet#804] [General]
Hi,

Is there a way to input a folder of images for ControlNet preprocessing instead of always giving it one image at a time manually?

What I want to do is to define a folder where ControlNet would look for images into the preprocessor and when hitting "Generate forever" in txt2img it would flip through the folder and input different image into preprocessing after each generated image

If one could just drag a folder into drop image field instead of a single image, it would be perfect...

**Comment by YourFriendlyNeighborhoodMONKE** [comment]
Nevermind, I found the solution after first not being able to find the right keywords:

https://github.com/Mikubill/sd-webui-controlnet/discussions/153

**Comment by ljleb** [comment]
Although you can do txt2img using the checkbox in the settings at the moment, #683 aims to provide better support for this.

**Comment by YourFriendlyNeighborhoodMONKE** [comment]
Wow, excellent work!

It's exactly how I would like it to work and hopefully it'll be merged very soon

When this is merged, does it mean that we'd be able to do this with img2img without changing those settings?

**Reply by ljleb** [reply]
Yes, it also works with simple img2img in the batch PR. You can do:
- (batch) controlnet + txt2img
- (batch) controlnet + img2img
- (batch) controlnet + batch img2img

**Reply by ljleb** [reply]
The way it works is, you can decide whether a controlnet model is in batch mode or not. As soon as one of the models is in batch mode, you enter batch mode for all active models.

**Reply by YourFriendlyNeighborhoodMONKE** [reply]
Excellent!

Thank you for your work

**Comment by YourFriendlyNeighborhoodMONKE** [comment]
@ljleb
Hey, I thought I might as well ask...

Would you be interested in making a function where batch processing would work with subfolder structures in a following way:

- I'd love to be able to split a folder into subfolders where each ControlNet preprocessing image would be in their own folder and then when I did inference with batch run with "Generate Forever" feature in the webui, the images it would generate would drop into the subfolder of the same controlnet image

This way when I did runs with wildcard prompts, it would be so much quicker to find all different versions of the images that are based on the same ControlNet source image instead of having to find them in one bigger folder

This could be just some kind of additional on/off function in the options

**Comment by bropines** [comment]
Here I am working with video, and I'm tired of running 1000 frames through the preprocessor every time. You can create a batch preprocessing interface. I know that all preprocessors are separate open source projects, but damn it, it’s easier to put them all into one of the interface tabs and stupidly either process one image or several, separately from the main interface.

In short, you need to preprocess files from folder to folder. So that later in the batch control net stupidly poke the path with ready-made frames
@ljleb

**Reply by ljleb** [reply]
We need to create a custom extras script for this. It's in my todo list, I just need to sort some things on my side for a while. If someone wants to create a custom postprocessing script using a1111's postprocessing api, that would be cool. Otherwise I'll end up writing the code eventually if no one looks into it.

**Reply by ljleb** [reply]
related: #1048
